cfy_manager/components/cluster/cluster.py

```diff
@@ -0,0 +1,187 @@
+            raise NetworkError(
+                'Could not connect to the REST service: '
+                '401 unauthorized. Possible access control misconfiguration'
+            )
+        if response.status_code != 200:
+            raise NetworkError(
+                'REST service returned an unexpected response: '
+                '{0}'.format(response.status_code)
+            )
+
+        try:
+            return json.loads(response.content)
+        except ValueError as e:
+            raise BootstrapError(
+                'REST service returned malformed JSON: {0}'.format(e))
+
+    def _get_join_addresses(self, master_manager_ip):
+        result = self._generic_cloudify_rest_request(
+            master_manager_ip,
+            80,
+            'v3.1/cluster/nodes',
+            'get')
+        return [n['host_ip'] for n in result['items']]
+
+    def _add_new_cluster_node_on_master(self, master_manager_ip,
+                                        cluster_node_host_ip,
+                                        cluster_node_name,
+                                        cluster_node_database_ip):
+        data = {
+            'host_ip': cluster_node_host_ip,
+            'node_name': cluster_node_name,
+            'database_ip': cluster_node_database_ip
+        }
+        result = self._generic_cloudify_rest_request(
+            master_manager_ip,
+            80,
+            'v3.1/cluster/nodes/{0}'.format(cluster_node_name),
+            'put',
+            data
+        )
+        return result
+
+    def _get_cluster_status(self, master_manager_ip):
+        result = self._generic_cloudify_rest_request(
+            master_manager_ip,
+            80,
+            'v3.1/cluster',
+            'get')
+        return result
+
+    def __generate_cluster_node_name(self):
+        chars = string.ascii_uppercase + string.digits
+        return 'cloudify_manager_' + ''.join(
+            random.choice(chars) for _ in range(NODE_NAME_GENERATED_CHAR_SIZE))
+
+    def _join_to_cluster(self, master_manager_ip):
+        cluster_node_host_ip = config[MANAGER][PUBLIC_IP]
+        cluster_node_name = \
+            config[CLUSTER][NODE_NAME] or self.__generate_cluster_node_name()
+        self.logger.notice('Adding cluster node "{0}" to the cluster'
+                           .format(cluster_node_name))
+        cluster_node_database_ip = \
+            config[POSTGRESQL_CLIENT]['host'].split(':')[0]
+        join_addrs = self._get_join_addresses(master_manager_ip)
+        new_node = self._add_new_cluster_node_on_master(
+            master_manager_ip, cluster_node_host_ip,
+            cluster_node_name, cluster_node_database_ip)
+        data = {
+            'host_ip': cluster_node_host_ip,
+            'node_name': cluster_node_name,
+            'credentials': new_node['credentials'],
+            'required': new_node['required'],
+            'join_addrs': join_addrs,
+            'ignore_database_validation': True,
```

Comment:
why this? and the `bootstrap_cluster` too

Reply:
They could have different values - E.G: starting a cluster with external database

Reply:
oh this is for the start_cluster_node, right